tests/usecases/list_permissions_test.py

```diff
@@ -0,0 +1,173 @@
+from datetime import datetime
+from typing import TYPE_CHECKING
+
+from grouper.constants import PERMISSION_CREATE
+from grouper.entities.pagination import PaginatedList, Pagination
+from grouper.entities.permission import Permission
+from grouper.graph import GroupGraph
+from grouper.models.base.constants import OBJ_TYPES
+from grouper.models.counter import Counter
+from grouper.models.group import Group
+from grouper.models.group_edge import GroupEdge
+from grouper.models.permission import Permission as SQLPermission
+from grouper.models.permission_map import PermissionMap
+from grouper.models.user import User
+from grouper.repositories.factory import RepositoryFactory
+from grouper.services.factory import ServiceFactory
+from grouper.usecases.factory import UseCaseFactory
+from grouper.usecases.list_permissions import ListPermissionsSortKey, ListPermissionsUI
+from tests.fixtures import session  # noqa: F401
+
+if TYPE_CHECKING:
+    from grouper.models.base.session import Session
+    from grouper.usecases.list_permissions import ListPermissions
+    from typing import Any, List, Optional
+
+
+class MockUI(ListPermissionsUI):
+    def __init__(self, sort=False):
+        # type: (bool) -> None
+        self.sort = sort
+
+    def listed_permissions(self, permissions, can_create):
+        # type: (PaginatedList[Permission], bool) -> None
+        self.permissions = permissions
+        if self.sort:
+            self.permissions.values = sorted(self.permissions.values)
+        self.can_create = can_create
+
+
+def assert_paginated_list_equal(left, right):
+    # type: (PaginatedList[Any], PaginatedList[Any]) -> None
+    """mypy NamedTuples don't compare with simple equality."""
+    assert left.values == right.values
+    assert left.total == right.total
+    assert left.offset == right.offset
+
+
+def create_test_data(session):  # noqa: F811
+    # type: (Session) -> List[Permission]
+    """Sets up a very basic test graph and returns the permission objects."""
+    early_date = datetime.utcfromtimestamp(1)
+    permissions = [
+        Permission(name="first-permission", description="first", created_on=datetime.utcnow()),
+        Permission(name="audited-permission", description="", created_on=datetime.utcnow()),
+        Permission(name="early-permission", description="is early", created_on=early_date),
+    ]
+    for permission in permissions:
+        sql_permission = SQLPermission(
```

Comment:
We should be able to replace this with calls to PermissionRepository when we refactor creating permissions, but I think it's already better than fixtures.

Reply:
Yup, agreed, and making it much nicer is what I'm going to work on today.